wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -0,0 +1,954 @@
+                                             .withMachineId(machine.getId())
+                                             .withWorkspaceId(workspaceId));
+
+                config = queuePeekOrFail(workspaceId);
+            } catch (RuntimeException | ServerException e) {
+                EnvironmentHolder env;
+                try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+                    env = environments.remove(workspaceId);
+                }
+                try {
+                    stopMachines(env.machines);
+                } catch (Exception remEx) {
+                    LOG.error(remEx.getLocalizedMessage(), remEx);
+                }
+                throw new ServerException(e.getLocalizedMessage(), e);
+            }
+        }
+    }
+
+    private void addMachine(String workspaceId,
+                            String machineId,
+                            String envName,
+                            String creator,
+                            MachineConfig machineConfig) throws ServerException {
+        Instance machine = new NoOpMachineInstance(MachineImpl.builder()
+                                                              .setConfig(machineConfig)
+                                                              .setId(machineId)
+                                                              .setWorkspaceId(workspaceId)
+                                                              .setStatus(MachineStatus.CREATING)
+                                                              .setEnvName(envName)
+                                                              .setOwner(creator)
+                                                              .build());
+        try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+            ensurePreDestroyIsNotExecuted();
+            EnvironmentHolder environmentHolder = environments.get(workspaceId);
+            if (environmentHolder != null && environmentHolder.status != EnvStatus.STOPPING) {
+                environmentHolder.machines.add(machine);
+            } else {
+                throw new ServerException(
+                        format("Can't add machine into environment. Environment of workspace '%s' is missing",
+                               workspaceId));
+            }
+        }
+    }
+
+    private void removeMachine(String workspaceId,
+                               String machineId) {
+        try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+            EnvironmentHolder environmentHolder = environments.get(workspaceId);
+            if (environmentHolder != null) {
+                for (int i = 0; i < environmentHolder.machines.size(); i++) {
+                    if (environmentHolder.machines.get(i).getId().equals(machineId)) {
+                        environmentHolder.machines.remove(i);
```

Comment:
Are you sure that it works fine? I mean when next use case:
```
i = 3
  1 2 3 4 5
      ^
```
you should remove i-th element and add 1 to i
```
i = 4
  1 2 4 5
        ^
```
Seems that you'll skip 4th element in this case.
Am I wrong?